graspologic/cluster/rclust.py

```diff
@@ -0,0 +1,377 @@
+# Copyright (c) Microsoft Corporation and contributors.
+# Licensed under the MIT License.
+
+import numpy as np
+from anytree import NodeMixin
+
+from .kclust import KMeansCluster
+from .autogmm import AutoGMMCluster
+from sklearn.utils.validation import check_is_fitted
+from sklearn.utils import check_array
+from sklearn.base import BaseEstimator
+from scipy.stats import chi2
+
+
+def _check_common_inputs(min_components, max_components, cluster_kws, embed_kws):
+    if not isinstance(min_components, int):
+        raise TypeError("min_components must be an int")
+    elif min_components < 1:
+        raise ValueError("min_components must be > 0")
+
+    if not isinstance(max_components, int):
+        raise TypeError("max_components must be an int")
+    elif max_components < 1:
+        raise ValueError("max_components must be > 0")
+    elif max_components < min_components:
+        raise ValueError("max_components must be >= min_components")
+
+    if embed_kws is not None:
+        if not isinstance(embed_kws, dict):
+            raise TypeError("embed_kws must be a dict")
+
+    if not isinstance(cluster_kws, dict):
+        raise TypeError("cluster_kws must be a dict")
+
+
+class RecursiveCluster(NodeMixin, BaseEstimator):
+    """
+    Recursively clusters based on a chosen clustering algorithm
+
+    Parameters
+    ----------
+    cluster_method : str {"GMM", "KMeans", "Spherical-KMeans"}, default="GMM"
+        The clustering method chosen to apply
+    min_components : int, optional (default=1)
+        The minimum number of mixture components or clusters to consider
+        for the first split
+    max_components : int, optional (default=10)
+        The maximum number of mixture components or clusters to consider
+        at each split
+    selection_criteria : str, optional, default(=None) is the default
```

Comment:
I actually took out `selection_criteria` cause I think users can just specify that in `cluster_kws` if needed. If they didn't specify "AIC" for "gmm", it would be "BIC"; while there is no choice for "kmeans"

Reply:
even better!